src/daft-plan/src/optimization/logical_plan_tracker.rs

```diff
@@ -0,0 +1,228 @@
+use std::{
+    collections::{hash_map::DefaultHasher, HashSet},
+    hash::{BuildHasher, BuildHasherDefault, Hash, Hasher},
+    num::NonZeroUsize,
+};
+
+use crate::LogicalPlan;
+
+/// A logical plan tracker that uses logical plan digests to detect optimization cycles.
+///
+/// The digests are cheaply hashable + comparable, and have the following guarantees:
+///
+///   p1 == p2 -> digest(p1) == digest(p2)
+///   hash(digest(p1)) == hash(digest(p2)) -> hash(p1) == hash(p2)
```

Comment:
I'm mainly trying to call out that `p1 == p2 -> digest(p1) == digest(p2) -> hash(digest(p1)) == hash(digest(p2))`, which allows us to put the digests in a hash set without worrying about false negatives (missing cycles), and hash collisions for the digests will be ~on par with hash collisions of the underlying logical plans (except a little bit better due to the inclusion of node counts).

Maybe the _modus tollens_ form would be more useful to give here, i.e. `hash(p1) != hash(p2) -> hash(digest(p1)) != hash(digest(p2))`.

Reply:
`hash(p1) != hash(p2) -> hash(digest(p1)) != hash(digest(p2))` is my way of talking about Case 2 that you gave, where the hash of the digest is stronger than the hash of just the plan due to the inclusion of the node count, which is good since we're using a relatively weak hashing mechanism for the plan; we are therefore making hash collisions less likely.

Reply:
@xcharleslin I just rebased onto most recent main (including your PR), and addressed some of your PR feedback in this commit: https://github.com/Eventual-Inc/Daft/pull/1292/commits/f2bdf564dbaafba71c9fa681d5a825527bee62b6

PTAL and let me know if I can improve the comments for the logical plan digest + tracker stuff any further!